File: patterns/agentic_rag/code/typescript/vercel-ai-sdk/agentic-rag.ts
/**
 * Agentic RAG — Vercel AI SDK variant.
 *
 * Pattern: Agentic RAG (hybrid retrieval + late reranking inside a reflection loop).
 * Framework: Vercel AI SDK (ai ^4.0.0) + @ai-sdk/anthropic (^1.0.0).
 * Idioms: dense and sparse search run in app code and are fused with
 *   reciprocal rank fusion; a reranker re-scores the fused candidates against
 *   the question; a sufficiency check retries with a widened query before
 *   generateText produces the cited answer.
 * Design doc: ../../../design.md (the framework-agnostic
 *   ../../python/agentic_rag.py walks the same hybrid retrieve, rerank,
 *   reflect, generate cycle with the same seed documents and query).
 *
 * Install:  pnpm add ai @ai-sdk/anthropic
 * Run:      ANTHROPIC_API_KEY=... npx tsx agentic-rag.ts
 *
 * The reranker here is a keyword-overlap stand-in so the file runs offline.
 * Replace it with a hosted reranker (Cohere Rerank) or a local cross-encoder
 * to ship; replace the in-memory hybrid index with a store that runs both
 * retrievals natively (Qdrant hybrid queries, or pgvector plus Postgres
 * full-text search).
 *
 * Note: ESM only (matches the canonical TypeScript project layout).
 */

import { createHash } from "node:crypto";

import { anthropic } from "@ai-sdk/anthropic";
import { generateText } from "ai";

// --- Types -------------------------------------------------------------------

interface Candidate {
  id: string;
  text: string;
  denseScore: number;
  sparseScore: number;
  fusedScore: number;
  rerankScore: number;
  metadata: Record<string, unknown>;
}

interface AgenticRAGResult {
  answer: string;
  evidence: Candidate[];
  attempts: number;
  grounded: boolean;
}

interface Embedder {
  embed(text: string): number[];
}

interface Reranker {
  rerank(question: string, candidates: Candidate[], topN: number): Candidate[];
}

// --- Deterministic mock embedder (matches the Python sibling) -----------------

class HashEmbedder implements Embedder {
  embed(text: string): number[] {
    const digest = createHash("md5").update(text).digest("hex");
    const big = BigInt(`0x${digest}`);
    const vec: number[] = [];
    for (let i = 0; i < 8; i++) {
      const shifted = (big >> BigInt(i * 8)) & 0xffn;
      vec.push(Number(shifted) / 255);
    }
    return vec;
  }
}

// --- Offline reranker stand-in -------------------------------------------------

class KeywordOverlapReranker implements Reranker {
  rerank(question: string, candidates: Candidate[], topN: number): Candidate[] {
    const terms = new Set(
      question
        .toLowerCase()
        .split(/\s+/)
        .filter((t) => t.length > 2),
    );
    for (const c of candidates) {
      const words = new Set(c.text.toLowerCase().split(/\s+/));
      let overlap = 0;
      for (const t of terms) if (words.has(t)) overlap++;
      c.rerankScore = terms.size ? overlap / terms.size : 0;
    }
    return [...candidates].sort((a, b) => b.rerankScore - a.rerankScore).slice(0, topN);
  }
}

// --- In-memory hybrid index ------------------------------------------------------

class HybridIndex {
  private items: Array<{ id: string; embedding: number[]; text: string; metadata: Record<string, unknown> }> = [];

  add(id: string, embedding: number[], text: string, metadata: Record<string, unknown>): void {
    this.items.push({ id, embedding, text, metadata });
  }

  private candidate(item: (typeof this.items)[number]): Candidate {
    return {
      id: item.id,
      text: item.text,
      denseScore: 0,
      sparseScore: 0,
      fusedScore: 0,
      rerankScore: 0,
      metadata: item.metadata,
    };
  }

  denseSearch(embedding: number[], topK: number): Candidate[] {
    const dot = (a: number[], b: number[]): number => a.reduce((sum, x, i) => sum + x * b[i], 0);
    return this.items
      .map((item) => ({ ...this.candidate(item), denseScore: dot(embedding, item.embedding) }))
      .sort((a, b) => b.denseScore - a.denseScore)
      .slice(0, topK);
  }

  sparseSearch(query: string, topK: number): Candidate[] {
    const terms = new Set(
      query
        .toLowerCase()
        .split(/\s+/)
        .filter((t) => t.length > 2),
    );
    return this.items
      .map((item) => {
        const words = new Set(item.text.toLowerCase().split(/\s+/));
        let overlap = 0;
        for (const t of terms) if (words.has(t)) overlap++;
        return { ...this.candidate(item), sparseScore: terms.size ? overlap / terms.size : 0 };
      })
      .sort((a, b) => b.sparseScore - a.sparseScore)
      .slice(0, topK);
  }
}

function reciprocalRankFusion(dense: Candidate[], sparse: Candidate[], k = 60): Candidate[] {
  // Summed reciprocal rank needs no score calibration, which is why RRF is
  // the default fusion for hybrid retrieval.
  const fused = new Map<string, Candidate>();
  for (const ranked of [dense, sparse]) {
    ranked.forEach((cand, rank) => {
      const slot = fused.get(cand.id) ?? cand;
      slot.fusedScore += 1 / (k + rank + 1);
      if (cand.denseScore) slot.denseScore = cand.denseScore;
      if (cand.sparseScore) slot.sparseScore = cand.sparseScore;
      fused.set(cand.id, slot);
    });
  }
  return [...fused.values()].sort((a, b) => b.fusedScore - a.fusedScore);
}

// --- Pipeline ---------------------------------------------------------------------

interface AgenticRAGOptions {
  retrieveK: number;
  rerankN: number;
  maxAttempts: number;
  sufficiencyThreshold: number;
}

class AgenticRAGPipeline {
  private index = new HybridIndex();

  constructor(
    private readonly embedder: Embedder,
    private readonly reranker: Reranker,
    private readonly model = anthropic("claude-haiku-4-5"),
    private readonly opts: AgenticRAGOptions = {
      retrieveK: 8,
      rerankN: 3,
      maxAttempts: 2,
      sufficiencyThreshold: 0.3,
    },
  ) {}

  ingest(documents: string[]): number {
    documents.forEach((doc, i) => {
      this.index.add(`doc_${i}`, this.embedder.embed(doc), doc, {});
    });
    return documents.length;
  }

  retrieve(question: string): { evidence: Candidate[]; attempts: number } {
    let current = question;
    let evidence: Candidate[] = [];
    for (let attempt = 1; attempt <= this.opts.maxAttempts; attempt++) {
      const dense = this.index.denseSearch(this.embedder.embed(current), this.opts.retrieveK);
      const sparse = this.index.sparseSearch(current, this.opts.retrieveK);
      const fused = reciprocalRankFusion(dense, sparse);
      evidence = this.reranker.rerank(question, fused, this.opts.rerankN);
      if (this.sufficient(evidence) || attempt === this.opts.maxAttempts) {
        return { evidence, attempts: attempt };
      }
      // Insufficient evidence: a production loop asks the sufficiency
      // reflector what is missing; the demo widens the query mechanically.
      current = `${question} background definition details`;
    }
    return { evidence, attempts: this.opts.maxAttempts };
  }

  private sufficient(evidence: Candidate[]): boolean {
    return evidence.length > 0 && evidence[0].rerankScore >= this.opts.sufficiencyThreshold;
  }

  async query(question: string): Promise<AgenticRAGResult> {
    const { evidence, attempts } = this.retrieve(question);
    const context = evidence.map((c, i) => `[${i + 1}] ${c.text}`).join("\n\n");
    const prompt =
      `Answer the question using only the numbered sources. Cite the source ` +
      `number for every claim. If the sources do not contain the answer, say ` +
      `so plainly.\n\nSources:\n${context}\n\nQuestion: ${question}`;

    const result = await generateText({ model: this.model, prompt });
    return { answer: result.text, evidence, attempts, grounded: this.sufficient(evidence) };
  }
}

// --- Smoke runner ---------------------------------------------------------------

const SEED_DOCUMENTS = [
  "ReAct is a prompting technique that combines chain-of-thought reasoning with action execution.",
  "RAG stands for Retrieval-Augmented Generation. It retrieves relevant documents before generating.",
  "Hybrid retrieval fuses dense embedding search with sparse keyword search using reciprocal rank fusion.",
  "A reranker re-scores retrieved candidates against the question before the model sees them.",
];

const QUESTION = "How does hybrid retrieval with reranking work?";

async function main(): Promise<void> {
  const pipeline = new AgenticRAGPipeline(new HashEmbedder(), new KeywordOverlapReranker());
  const n = pipeline.ingest(SEED_DOCUMENTS);
  console.log(`Ingested ${n} documents`);

  if (!process.env.ANTHROPIC_API_KEY) {
    console.log("Skipping generate — set ANTHROPIC_API_KEY to exercise the real model call.");
    const { evidence, attempts } = pipeline.retrieve(QUESTION);
    console.log(`Evidence: ${evidence.length} chunks after ${attempts} attempt(s) (offline).`);
    for (const c of evidence) {
      console.log(`  fused=${c.fusedScore.toFixed(4)} rerank=${c.rerankScore.toFixed(2)} ${c.text.slice(0, 60)}`);
    }
    return;
  }

  const result = await pipeline.query(QUESTION);
  console.log(`Evidence: ${result.evidence.length} chunks after ${result.attempts} attempt(s)`);
  console.log(`Answer: ${result.answer.slice(0, 120)}`);
}

// ESM entry-point check.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

export { AgenticRAGPipeline, HybridIndex, HashEmbedder, KeywordOverlapReranker, reciprocalRankFusion };
export type { Candidate, AgenticRAGResult, Embedder, Reranker, AgenticRAGOptions };

File: patterns/agentic_rag/code/python/agentic_rag.py
"""
Agentic RAG — hybrid retrieval with late reranking inside a reflection loop.

The retrieval core that separates this pattern from baseline RAG:
  Hybrid retrieve: dense (embedding) and sparse (keyword) search run in
                   parallel, fused with reciprocal rank fusion (RRF)
  Rerank:          a late reranker re-scores the fused candidates against
                   the question between retrieval and generation
  Reflect:         a sufficiency check decides answer vs retry with a
                   reformulated query (capped attempts)

Design doc:  ../../design.md
Overview:    ../../overview.md
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol

from patterns.agentic_rag.schemas.state import (  # noqa: F401
    AgenticRagState,
    EvidenceChunk,
    SubQuestion,
)

# --- Interfaces --------------------------------------------------------------
#
# Recipes targeting Agentic RAG bind to the canonical ``AgenticRagState`` /
# ``SubQuestion`` / ``EvidenceChunk`` shapes imported above (EvidenceChunk
# carries ``embedding_score`` and ``rerank_score`` for the two stages below).
# This sibling's ``Candidate`` dataclass is the in-memory retrieval container;
# an adapter pairs the two at the evidence boundary.


class LLM(Protocol):
    def generate(self, messages: list[dict]) -> str: ...


class Embedder(Protocol):
    """Convert text to a float vector. Use any embedding model."""

    def embed(self, text: str) -> list[float]: ...


class Reranker(Protocol):
    """Re-score candidates against the question with full attention.

    Replace with a hosted reranker (Cohere Rerank) or a local cross-encoder
    to ship; the keyword-overlap mock in the demo keeps this file offline.
    """

    def rerank(self, question: str, candidates: list[Candidate], top_n: int) -> list[Candidate]: ...


# --- Core types ---------------------------------------------------------------


@dataclass
class Candidate:
    id: str
    text: str
    dense_score: float = 0.0
    sparse_score: float = 0.0
    fused_score: float = 0.0
    rerank_score: float = 0.0
    metadata: dict = field(default_factory=dict)


@dataclass
class AgenticRAGResult:
    answer: str
    evidence: list[Candidate] = field(default_factory=list)
    attempts: int = 1
    grounded: bool = True


# --- In-memory hybrid index ----------------------------------------------------


class HybridIndex:
    """Dense dot-product plus keyword-overlap search over one corpus.

    Replace with a store that runs both retrievals natively to ship —
    Qdrant hybrid queries, or pgvector alongside Postgres full-text search.
    """

    def __init__(self) -> None:
        self._items: list[tuple[str, list[float], str, dict]] = []

    def add(self, id: str, embedding: list[float], text: str, metadata: dict) -> None:
        self._items.append((id, embedding, text, metadata))

    def dense_search(self, embedding: list[float], top_k: int) -> list[Candidate]:
        def dot(a: list[float], b: list[float]) -> float:
            return sum(x * y for x, y in zip(a, b, strict=False))

        scored = [
            Candidate(id=id_, text=text, dense_score=dot(embedding, emb), metadata=meta)
            for id_, emb, text, meta in self._items
        ]
        scored.sort(key=lambda c: c.dense_score, reverse=True)
        return scored[:top_k]

    def sparse_search(self, query: str, top_k: int) -> list[Candidate]:
        terms = {t for t in query.lower().split() if len(t) > 2}
        scored = []
        for id_, _emb, text, meta in self._items:
            words = set(text.lower().split())
            overlap = len(terms & words) / len(terms) if terms else 0.0
            scored.append(Candidate(id=id_, text=text, sparse_score=overlap, metadata=meta))
        scored.sort(key=lambda c: c.sparse_score, reverse=True)
        return scored[:top_k]


def reciprocal_rank_fusion(dense: list[Candidate], sparse: list[Candidate], k: int = 60) -> list[Candidate]:
    """Fuse two ranked lists by summed reciprocal rank — no score calibration
    needed, which is why RRF is the default fusion for hybrid retrieval."""
    fused: dict[str, Candidate] = {}
    for ranked in (dense, sparse):
        for rank, cand in enumerate(ranked):
            slot = fused.setdefault(cand.id, cand)
            slot.fused_score += 1.0 / (k + rank + 1)
            if cand.dense_score:
                slot.dense_score = cand.dense_score
            if cand.sparse_score:
                slot.sparse_score = cand.sparse_score
    return sorted(fused.values(), key=lambda c: c.fused_score, reverse=True)


# --- Pipeline -------------------------------------------------------------------

ANSWER_PROMPT = """\
Answer the question using only the numbered sources. Cite the source number
for every claim. If the sources do not contain the answer, say so plainly.

Sources:
{context}

Question: {question}"""


class AgenticRAGPipeline:
    """Hybrid retrieve, rerank, reflect, then generate with citations.

    Ingestion: call pipeline.ingest(documents) once (or incrementally).
    Query:     call pipeline.query(question) at request time.
    """

    def __init__(
        self,
        llm: LLM,
        embedder: Embedder,
        reranker: Reranker,
        index: HybridIndex | None = None,
        retrieve_k: int = 8,
        rerank_n: int = 3,
        max_attempts: int = 2,
        sufficiency_threshold: float = 0.3,
    ):
        self.llm = llm
        self.embedder = embedder
        self.reranker = reranker
        self.index = index or HybridIndex()
        self.retrieve_k = retrieve_k
        self.rerank_n = rerank_n
        self.max_attempts = max_attempts
        self.sufficiency_threshold = sufficiency_threshold

    def ingest(self, documents: list[str], metadata: list[dict] | None = None) -> int:
        metadata = metadata or [{}] * len(documents)
        for i, (doc, meta) in enumerate(zip(documents, metadata, strict=False)):
            self.index.add(f"doc_{i}", self.embedder.embed(doc), doc, meta)
        return len(documents)

    def query(self, question: str) -> AgenticRAGResult:
        current = question
        evidence: list[Candidate] = []
        for attempt in range(1, self.max_attempts + 1):
            dense = self.index.dense_search(self.embedder.embed(current), self.retrieve_k)
            sparse = self.index.sparse_search(current, self.retrieve_k)
            fused = reciprocal_rank_fusion(dense, sparse)
            evidence = self.reranker.rerank(question, fused, self.rerank_n)
            if self._sufficient(evidence) or attempt == self.max_attempts:
                return self._compose(question, evidence, attempt)
            # Insufficient evidence: reformulate and retry. A production loop
            # asks the sufficiency reflector (prompts/sufficiency-reflector.md)
            # what is missing; the demo widens the query mechanically.
            current = f"{question} background definition details"
        return self._compose(question, evidence, self.max_attempts)

    def _sufficient(self, evidence: list[Candidate]) -> bool:
        return bool(evidence) and evidence[0].rerank_score >= self.sufficiency_threshold

    def _compose(self, question: str, evidence: list[Candidate], attempts: int) -> AgenticRAGResult:
        context = "\n\n".join(f"[{i + 1}] {c.text}" for i, c in enumerate(evidence))
        answer = self.llm.generate(
            [{"role": "user", "content": ANSWER_PROMPT.format(context=context, question=question)}]
        )
        return AgenticRAGResult(
            answer=answer,
            evidence=evidence,
            attempts=attempts,
            grounded=self._sufficient(evidence),
        )


# --- Example ---------------------------------------------------------------------

if __name__ == "__main__":
    import hashlib

    class MockEmbedder:
        """Deterministic fake embeddings based on text hash."""

        def embed(self, text: str) -> list[float]:
            h = int(hashlib.md5(text.encode()).hexdigest(), 16)
            return [(h >> i & 0xFF) / 255.0 for i in range(8)]

    class KeywordOverlapReranker:
        """Offline stand-in for a cross-encoder: question-term overlap."""

        def rerank(self, question: str, candidates: list[Candidate], top_n: int) -> list[Candidate]:
            terms = {t for t in question.lower().split() if len(t) > 2}
            for c in candidates:
                words = set(c.text.lower().split())
                c.rerank_score = len(terms & words) / len(terms) if terms else 0.0
            return sorted(candidates, key=lambda c: c.rerank_score, reverse=True)[:top_n]

    class MockLLM:
        def generate(self, messages: list[dict]) -> str:
            return f"Grounded answer from: {messages[-1]['content'][:80]}..."

    pipeline = AgenticRAGPipeline(
        llm=MockLLM(),
        embedder=MockEmbedder(),
        reranker=KeywordOverlapReranker(),
    )
    n = pipeline.ingest(
        [
            "ReAct is a prompting technique that combines chain-of-thought reasoning with action execution.",
            "RAG stands for Retrieval-Augmented Generation. It retrieves relevant documents before generating.",
            "Hybrid retrieval fuses dense embedding search with sparse keyword search using reciprocal rank fusion.",
            "A reranker re-scores retrieved candidates against the question before the model sees them.",
        ]
    )
    print(f"Ingested {n} documents")

    result = pipeline.query("How does hybrid retrieval with reranking work?")
    print(f"Evidence: {len(result.evidence)} chunks after {result.attempts} attempt(s)")
    for i, c in enumerate(result.evidence):
        print(f"  [{i + 1}] fused={c.fused_score:.4f} rerank={c.rerank_score:.2f} {c.text[:60]}")
    print(f"Answer: {result.answer[:120]}")
